packages/ethereum/contracts/src/Crypto.sol

```diff
@@ -0,0 +1,598 @@
+    return address(uint160(uint256(keccak256(abi.encodePacked(p.x, p.y)))));
+  }
+
+  /**
+   * Adds two elliptic curve points using the general implementation.
+   *
+   * This function is optimized to perform one point addition.
+   *
+   * Optimizations:
+   * - solidity assembly
+   * - optimize for a single addition
+   */
+  function ecAdd(CurvePoint memory p, CurvePoint memory q) public view returns (CurvePoint memory r)  {
+    if (p.x == 0 && p.y == 0) { // Inf + Q = Q
+      return q;
+    }
+
+    if (q.x == 0 && q.y == 0) { // P + Inf = P
+      return p;
+    }
+
+    assembly {
+      let toInvert := addmod( // q.x - p.x
+        mload(q), // q.x 
+        sub(SECP256K1_BASE_FIELD_ORDER, mload(p)), // - p.x
+        SECP256K1_BASE_FIELD_ORDER
+      )
+
+      let payload := mload(0x40)
+      mstore(payload, 0x20)             // Length of Base
+      mstore(add(payload, 0x20), 0x20)  // Length of Exponent
+      mstore(add(payload, 0x40), 0x20)  // Length of Modulus
+      mstore(add(payload, 0x60), toInvert)  // Base
+      mstore(add(payload, 0x80), 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEFFFFFC2D) // p - 1
+      mstore(add(payload, 0xa0), SECP256K1_BASE_FIELD_ORDER)     // Modulus
+      if iszero(staticcall(not(0), 0x05, payload, 0xC0, payload, 0x20)) {
+        revert(0, 0)
+      }
+
+      let lambda := mulmod( // (q.y - p.y) * (q.x - p.x) ^ -1
+        addmod( // q.y - p.y
+          mload(add(q, 0x20)), // q.y
+          sub(SECP256K1_BASE_FIELD_ORDER, mload(add(p, 0x20))), // - p.y
+          SECP256K1_BASE_FIELD_ORDER
+        ), 
+        mload(payload), // (q.x - p.x) ^ -1
+        SECP256K1_BASE_FIELD_ORDER
+      )
+
+      mstore(
+        r,
+        addmod( // lambda^2 - q.x - p.x
+          mulmod(lambda, lambda, SECP256K1_BASE_FIELD_ORDER), // lambda^2
+          addmod( // - q.x - p.x
+            sub(SECP256K1_BASE_FIELD_ORDER, mload(q)), // - q.x 
+            sub(SECP256K1_BASE_FIELD_ORDER, mload(p)), // - p.x
+            SECP256K1_BASE_FIELD_ORDER
+          ),
+          SECP256K1_BASE_FIELD_ORDER
+        )
+      )
+
+      mstore(
+        add(r, 0x20), // r.y
+        addmod( // lambda * (p.x - r.x) - p.y
+          mulmod( // lambda * (p.x - r.x)
+            lambda, 
+            addmod( // p.x - r.x
+              mload(p), // p.x 
+              sub(SECP256K1_BASE_FIELD_ORDER, mload(r)), // - r.x
+              SECP256K1_BASE_FIELD_ORDER
+            ), 
+            SECP256K1_BASE_FIELD_ORDER
+          ),
+          sub(SECP256K1_BASE_FIELD_ORDER, mload(add(p, 0x20))), 
+          SECP256K1_BASE_FIELD_ORDER
+        )
+      )
+    }
+  }
+
+  /**
+   * Consumes a byte string and returns a sep256k1 curvepoint.
+   * 
+   * Implements secp256k1_XMD:SHA-256_SSWU_RO_, see
+   * https://www.ietf.org/archive/id/draft-irtf-cfrg-hash-to-curve-16.html
+   */
+  function hashToCurve(bytes memory payload, bytes memory DST) public view returns (CurvePoint memory) {
```

Comment:
The domain separation tag should also contain information about our used algorithm, that is `secp256k1_XMD:SHA-256_SSWU_RO_`

Reply:
The idea is to have sth like `hopr-redeemTicket-v2.0` - the chosen hash_to_curve should be then implicitly given